GG/src/Font.cpp

```diff
@@ -685,33 +700,167 @@ Font::Glyph::Glyph(const boost::shared_ptr<Texture>& texture, const Pt& ul, cons
+            bool retval = m_tag_stack.empty() ? false : m_tag_stack.top() == sub;
+            if (retval) {
+                m_tag_stack.pop();
+                if (m_tag_stack.empty() || m_tag_stack.top() != PRE_TAG)
+                    m_ignore_tags = false;
+            }
+            return retval;
+        }
+
+        const std::string* m_text;
+        const std::set<std::string>* m_known_tags;
+        bool m_ignore_tags;
+
+        // m_tag_stack is used to track XML opening/closing tags.
+        std::stack<Font::Substring> m_tag_stack;
+
+        // The combined regular expression.
+        xpr::sregex m_EVERYTHING;
+    };
+
+    class TagHandler {
+        public:
+        TagHandler() :
+            m_known_tags(),
+            m_regex_w_tags(m_known_tags, false),
+            m_regex_w_tags_skipping_unmatched(m_known_tags, true)
+        {}
+
+        void Insert(const std::string& tag) {
+            m_known_tags.insert(tag);
+        }
+
+        void Erase(const std::string& tag) {
+            m_known_tags.erase(tag);
+        }
+
+        void Clear() {
+            m_known_tags.clear();
+        }
+
+        const std::set<std::string>& KnownTags()
+        { return m_known_tags; }
+        // Return a regex bound to \p text using the currently known
+        // tags possible \p ignore_tags and/or \p strip_unpaired_tags
+        xpr::sregex & Regex(const std::string& text, bool ignore_tags, bool strip_unpaired_tags = false) {
+            if (!strip_unpaired_tags)
+                return m_regex_w_tags.BindRegexToText(text, ignore_tags);
+            else
+                return m_regex_w_tags_skipping_unmatched.BindRegexToText(text, ignore_tags);
+        }
+
+        private:
+        // set of tags known to the handler
+        std::set<std::string> m_known_tags;
+
+        // Compiled regular expression including tag stack
+        CompiledRegex m_regex_w_tags;
+
+        // Compiled regular expression using tags but skipping unmatched tags.
+        CompiledRegex m_regex_w_tags_skipping_unmatched;
+    };
+
 
     // Global set of known tags. Wrapped in a function for deterministic static initialization order.
-    std::set<std::string>& KnownTags()
+    TagHandler& StaticTagHandler()
     {
-        static std::set<std::string> known_tags;
-        return known_tags;
+        static TagHandler tag_handler;
+        return tag_handler;
     }
 
     // Registers the default action and known tags.
     int RegisterDefaultTags() {
-        ActionTags().insert("i");
-        ActionTags().insert("s");
-        ActionTags().insert("u");
-        ActionTags().insert("rgba");
-        ActionTags().insert(ALIGN_LEFT_TAG);
-        ActionTags().insert(ALIGN_CENTER_TAG);
-        ActionTags().insert(ALIGN_RIGHT_TAG);
-        ActionTags().insert(PRE_TAG);
-
-        // Always know the action tags.
-        KnownTags().insert(ActionTags().begin(), ActionTags().end());
+        StaticTagHandler().Insert("i");
```

Comment:
This stems from old code but while you're fixing this already: the literals should be replaced with constants like it was already done for ALIGN_LEFT_TAG and relatives.